ams/server/src/main/java/com/netease/arctic/server/optimizing/plan/CommonPartitionEvaluator.java

```diff
@@ -105,36 +108,66 @@ private boolean isDuplicateDelete(IcebergContentFile<?> delete) {
     return deleteExist;
   }
 
-  private void addFragmentFile(IcebergDataFile dataFile, List<IcebergContentFile<?>> deletes) {
+  private boolean addFragmentFile(IcebergDataFile dataFile, List<IcebergContentFile<?>> deletes) {
+    if (!fileShouldRewrite(dataFile, deletes)) {
+      return false;
+    }
     fragmentFileSize += dataFile.fileSizeInBytes();
     fragmentFileCount += 1;
 
     for (IcebergContentFile<?> delete : deletes) {
       addDelete(delete);
     }
+    return true;
   }
 
-  private void addSegmentFile(IcebergDataFile dataFile, List<IcebergContentFile<?>> deletes) {
-    segmentFileSize += dataFile.fileSizeInBytes();
-    segmentFileCount += 1;
-
-    if (shouldRewriteSegmentFile(dataFile, deletes)) {
+  private boolean addSegmentFile(IcebergDataFile dataFile, List<IcebergContentFile<?>> deletes) {
+    if (fileShouldRewrite(dataFile, deletes)) {
       rewriteSegmentFileSize += dataFile.fileSizeInBytes();
       rewriteSegmentFileCount += 1;
-    } else if (shouldRewritePosForSegmentFile(dataFile, deletes)) {
+    } else if (segmentFileShouldRewritePos(dataFile, deletes)) {
       rewritePosSegmentFileSize += dataFile.fileSizeInBytes();
       rewritePosSegmentFileCount += 1;
+    } else {
+      return false;
     }
+
+    segmentFileSize += dataFile.fileSizeInBytes();
+    segmentFileCount += 1;
     for (IcebergContentFile<?> delete : deletes) {
       addDelete(delete);
     }
+    return true;
   }
 
-  public boolean shouldRewriteSegmentFile(IcebergDataFile dataFile, List<IcebergContentFile<?>> deletes) {
+  protected boolean fileShouldFullOptimizing(IcebergDataFile dataFile, List<IcebergContentFile<?>> deleteFiles) {
+    if (config.isFullRewriteAllFiles()) {
+      return true;
+    }
+    if (isFragmentFile(dataFile)) {
+      return true;
+    }
+    // if a file is related any delete files or is not big enough, it should full optimizing
+    return !deleteFiles.isEmpty() || dataFile.fileSizeInBytes() < config.getTargetSize() * 0.9;
+  }
+
+  public boolean fileShouldRewrite(IcebergDataFile dataFile, List<IcebergContentFile<?>> deletes) {
+    if (isFullOptimizing()) {
+      return fileShouldFullOptimizing(dataFile, deletes);
+    }
+    if (isFragmentFile(dataFile)) {
+      return true;
+    }
     return getRecordCount(deletes) > dataFile.recordCount() * config.getMajorDuplicateRatio();
   }
 
-  public boolean shouldRewritePosForSegmentFile(IcebergDataFile dataFile, List<IcebergContentFile<?>> deletes) {
+  public boolean segmentFileShouldRewritePos(IcebergDataFile dataFile, List<IcebergContentFile<?>> deletes) {
+    if (isFullOptimizing()) {
+      return false;
+    }
+    if (isFragmentFile(dataFile)) {
+      return false;
+    }
     if (deletes.stream().anyMatch(delete -> delete.content() == FileContent.EQUALITY_DELETES)) {
       return true;
     } else if (deletes.stream().filter(delete -> delete.content() == FileContent.POSITION_DELETES).count() >= 2) {
```

Comment:
Why we need at least 2 position deletes?

Reply:
The purpose here is to keep the segment file unchanged and generate a pos-delete for it. If there is only one pos- delete file, there is no need to regenerate it.